src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1426,10 +1426,81 @@ public static unsafe nuint NarrowUtf16ToAscii(char* pUtf16Buffer, byte* pAsciiBu
             goto Finish;
         }
 
-        private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool VectorContainsNonAsciiChar(Vector128<ushort> utf16Vector, Vector128<short> asciiMaskForTestZ, Vector128<ushort> asciiMaskForAddSaturate)
         {
-            // This method contains logic optimized for both SSE2 and SSE41. Much of the logic in this method
-            // will be elided by JIT once we determine which specific ISAs we support.
+            if (Sse2.IsSupported)
+            {
+                // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
+
+                if (Sse41.IsSupported)
+                {
+                    if (!Sse41.TestZ(utf16Vector.AsInt16(), asciiMaskForTestZ))
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16Vector, asciiMaskForAddSaturate).AsByte()) & 0b_1010_1010_1010_1010) != 0)
+                    {
+                        return true;
+                    }
+                }
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                Vector128<ushort> maxChars = AdvSimd.Arm64.MaxPairwise(utf16Vector, utf16Vector);
+                if ((maxChars.AsUInt64().ToScalar() & 0xFF80FF80FF80FF80) != 0)
+                {
+                    return true; // found non-ASCII data
+                }
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<byte> ExtractAsciiVector(Vector128<ushort> vectorFirst, Vector128<ushort> vectorSecond)
+        {
+            if (Sse2.IsSupported)
+            {
+                return Sse2.PackUnsignedSaturate(vectorFirst.AsInt16(), vectorSecond.AsInt16());
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                return AdvSimd.Arm64.UnzipEven(vectorFirst.AsByte(), vectorSecond.AsByte());
```

Comment:
Its a little bit unfortunate that we're using `UnzipEven` here but an entirely different path of `PackUnsignedSaturate` on x86/x64.

Conceivably x86/x64 could be `UnpackLow` and the logic could be nearly identical. That would make it simpler, long term, to normalize to `Shuffle(x, y, cns_mask)`

Reply:
True. Maybe the `Vector128.Shuffle()` can hide these differences in the future.